jupyter_server_proxy/websockify.py

```diff
@@ -0,0 +1,88 @@
+"""
+A simple translation layer between tornado websockets and asyncio stream
+connections.
+
+This subsumes the functionality of websockify
+(https://github.com/novnc/websockify) without needing an extra proxy hop
+or process through with all messages pass for translation.
+"""
+
+import asyncio
+
+from .handlers import NamedLocalProxyHandler, SuperviseAndProxyHandler
+
+class WebsockifyProtocol(asyncio.Protocol):
+    """
+    A protocol handler for the proxied stream connection.
+    Sends any received blocks directly as websocket messages.
+    """
+    def __init__(self, handler):
+        self.handler = handler
+
+    def data_received(self, data):
+        "Send the buffer as a websocket message."
+        self.handler._record_activity()
+        self.handler.write_message(data, binary=True) # async, no wait
```

Comment:
Looking at https://www.tornadoweb.org/en/stable/websocket.html#tornado.websocket.WebSocketHandler.write_message, I see that it does return a `Future`. Can you help me understand why this doesn't need to be awaited here? I see that `data_received` is not a coroutine, and Guido explains [why](https://groups.google.com/g/python-tulip/c/z-IVH5RoDzo/m/SpZc0zTuPJsJ). But I don't fully understand what happens here if `data_received` is called twice with data (in order), but the previous `write_message` has not completed (because `write_message` is async). Would this not garble the data and have it be transmited out of order? Glyph's comment further down that thread has the same question as me...

If you think this isn't going to cause ordering problems, adding a comment explaining why would be very helpful!

Reply:
This also threw me off but digging into tornado a bit I'm pretty sure it's okay. `write_message` queues up the message before doing anything async, and the async returned is meant only to be used (optionally) for control flow, so ordering is preserved, even if the first messages hasn't been sent. There's a bit of commentary on this [here](https://github.com/tornadoweb/tornado/blob/bdfc017c66817359158185561cee7878680cd841/tornado/websocket.py#L1073). There may be an issue determining exactly which messages have been acknowledged if the connection is closed or there's another error, but given that that will abort the whole thing, and there's no way to communicate that anyway, that should be fine. I'll try to clarify this in comments.

Reply:
https://github.com/tornadoweb/tornado/blob/bdfc017c66817359158185561cee7878680cd841/tornado/websocket.py#L1073 is perfect, include it in the comment as well and I think we're good.